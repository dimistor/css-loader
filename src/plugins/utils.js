import postcss from 'postcss';

const createICSSImports = (imports) =>
  Object.keys(imports).map((path) => {
    const aliases = imports[path];
    const declarations = Object.keys(aliases).map((key) =>
      postcss.decl({
        prop: key,
        value: aliases[key],
        raws: { before: '\n  ' },
      })
    );

    return postcss
      .rule({
        selector: `:import(${path})`,
        raws: { after: '\n' },
      })
      .append(declarations);
  });

export default { createICSSImports };
