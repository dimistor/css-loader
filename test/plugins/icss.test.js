import fs from 'fs';
import path from 'path';

import plugin from '../../src/plugins/icss';
import { runPostcss } from '../helpers/utils';

describe('ICSS postcss plugin', () => {
  it('basic', async () => {
    const result = await runPostcss(
      fs.readFileSync(
        path.join(__dirname, '../fixtures/icss/import-export.css')
      ),
      [plugin]
    );

    expect(result.css).toMatchSnapshot('css');
    expect(result.messages).toMatchSnapshot('messages');
  });

  it('only import messages', async () => {
    const result = await runPostcss(
      fs.readFileSync(path.join(__dirname, '../fixtures/icss/only-import.css')),
      [plugin]
    );

    expect(result.css).toMatchSnapshot('css');
    expect(result.messages).toMatchSnapshot('messages');
  });

  it('only export messages', async () => {
    const result = await runPostcss(
      fs.readFileSync(path.join(__dirname, '../fixtures/icss/only-export.css')),
      [plugin]
    );

    expect(result.css).toMatchSnapshot('css');
    expect(result.messages).toMatchSnapshot('messages');
  });
});
