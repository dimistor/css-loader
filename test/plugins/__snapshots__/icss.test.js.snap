// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ICSS postcss plugin basic: css 1`] = `
":import() {
}

.foo {
  color: red;
}

@media only screen and (max-width: 600px) {
  body {
    background-color: lightblue;
  }
}

:export() {
  foo: bar
}
"
`;

exports[`ICSS postcss plugin basic: messages 1`] = `
Array [
  Object {
    "plugin": "postcss-css-loader-icss",
    "tokens": Object {
      "duplicate.css": Object {
        "bar": "foo",
        "color": "green",
        "foo": "bar",
      },
      "http://example.com/other-style.css": Object {
        "{media}": "screen and print",
      },
      "http://example.com/style.css": Object {},
      "other-test.css": Object {
        "key": "value",
        "other-key": "other-value",
      },
      "test-media.css": Object {
        "{media}": "screen and print",
      },
      "test.css": Object {},
    },
    "type": "import",
  },
  Object {
    "plugin": "postcss-css-loader-icss",
    "tokens": Object {
      "bar": "foo",
      "foo": "bar",
    },
    "type": "export",
  },
]
`;

exports[`ICSS postcss plugin only export messages: css 1`] = `
"
"
`;

exports[`ICSS postcss plugin only export messages: messages 1`] = `
Array [
  Object {
    "plugin": "postcss-css-loader-icss",
    "tokens": Object {
      "foo": "bar",
    },
    "type": "export",
  },
]
`;

exports[`ICSS postcss plugin only import messages: css 1`] = `
"
"
`;

exports[`ICSS postcss plugin only import messages: messages 1`] = `
Array [
  Object {
    "plugin": "postcss-css-loader-icss",
    "tokens": Object {
      "img.png": Object {
        "__url_0": "default",
      },
    },
    "type": "import",
  },
]
`;
